internal/database/sqlcommon/tokentransfer_sql.go

```diff
@@ -69,86 +70,89 @@ var (
+		}, requestConflictEmptyResult)
+	return err
+}
+
+func (s *SQLCommon) InsertOrGetTokenTransfer(ctx context.Context, transfer *core.TokenTransfer) (existing *core.TokenTransfer, err error) {
 	ctx, tx, autoCommit, err := s.BeginOrUseTx(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer s.RollbackTx(ctx, tx, autoCommit)
 
-	rows, _, err := s.QueryTx(ctx, tokentransferTable, tx,
-		sq.Select("seq").
-			From(tokentransferTable).
-			Where(sq.Eq{
-				"protocol_id": transfer.ProtocolID,
-				"namespace":   transfer.Namespace,
-			}),
-	)
-	if err != nil {
-		return err
-	}
-	existing := rows.Next()
-	rows.Close()
-
-	if existing {
-		if _, err = s.UpdateTx(ctx, tokentransferTable, tx,
-			sq.Update(tokentransferTable).
-				Set("type", transfer.Type).
-				Set("local_id", transfer.LocalID).
-				Set("pool_id", transfer.Pool).
-				Set("token_index", transfer.TokenIndex).
-				Set("uri", transfer.URI).
-				Set("connector", transfer.Connector).
-				Set("key", transfer.Key).
-				Set("from_key", transfer.From).
-				Set("to_key", transfer.To).
-				Set("amount", transfer.Amount).
-				Set("message_id", transfer.Message).
-				Set("message_hash", transfer.MessageHash).
-				Set("tx_type", transfer.TX.Type).
-				Set("tx_id", transfer.TX.ID).
-				Set("blockchain_event", transfer.BlockchainEvent).
-				Where(sq.Eq{"protocol_id": transfer.ProtocolID}),
-			func() {
-				s.callbacks.UUIDCollectionNSEvent(database.CollectionTokenTransfers, core.ChangeEventTypeUpdated, transfer.Namespace, transfer.LocalID)
-			},
-		); err != nil {
-			return err
-		}
-	} else {
+	if transfer.Created == nil {
 		transfer.Created = fftypes.Now()
-		if _, err = s.InsertTx(ctx, tokentransferTable, tx,
-			sq.Insert(tokentransferTable).
-				Columns(tokenTransferColumns...).
-				Values(
-					transfer.Type,
-					transfer.LocalID,
-					transfer.Pool,
-					transfer.TokenIndex,
-					transfer.URI,
-					transfer.Connector,
-					transfer.Namespace,
-					transfer.Key,
-					transfer.From,
-					transfer.To,
-					transfer.Amount,
-					transfer.ProtocolID,
-					transfer.Message,
-					transfer.MessageHash,
-					transfer.TX.Type,
-					transfer.TX.ID,
-					transfer.BlockchainEvent,
-					transfer.Created,
-				),
-			func() {
-				s.callbacks.UUIDCollectionNSEvent(database.CollectionTokenTransfers, core.ChangeEventTypeCreated, transfer.Namespace, transfer.LocalID)
-			},
-		); err != nil {
-			return err
-		}
 	}
+	opErr := s.attemptTokenTransferEventInsert(ctx, tx, transfer, true /* we want a failure here we can progress past */)
+	if opErr == nil {
+		return nil, s.CommitTx(ctx, tx, autoCommit)
+	}
+
+	// Do a select within the transaction to determine if the protocolID already exists
+	existing, err = s.GetTokenTransferByProtocolID(ctx, transfer.Namespace, transfer.Connector, transfer.ProtocolID)
+	if err != nil || existing != nil {
+
+		if existing != nil {
+			if _, err = s.UpdateTx(ctx, tokentransferTable, tx,
+				sq.Update(tokentransferTable).
```

Comment:
Did you determine if there's a situation where we should update an existing transfer? My thinking was that this shouldn't actually happen (and any code previously added for updates may need to be removed).

Reply:
I'm not sure I'm 100% confident that we never update them, but I can't think of a reason. I've added commit 0ebd4adae8e73b89150bec35a398c864ffac81a3 to remove the update logic. I think I could do with your guidance if any more tidy up needs doing that might have been assuming token transfers can be updated?